forte/data/extractor/extractor.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import codecs
+import logging
+import os
+import numpy as np
+import torch
+from torch import Tensor
+from typing import Iterator, Dict, List, Any, Union, Set, Tuple
+
+from forte.data.data_pack import DataPack
+from forte.data.data_utils_io import dataset_path_iterator
+from forte.data.readers.base_reader import PackReader
+from ft.onto.base_ontology import Token, Sentence, Document, Annotation, EntityMention
+from forte.data.ontology.core import EntryType
+from forte.data.span import Span
+from forte.data.extractor.vocabulary import Vocabulary
+from forte.data.extractor.feature import Feature
+
+
+class BaseExtractor:
+    def __init__(self, config: Dict):
+        self.config = config
+        self.entry = config["entry"]
+        use_pad = config.get("vocab_use_pad", True)
+        use_unk = config.get("vocab_use_unk", False)
+        method = config.get("vocab_method", "indexing")
+        self.__vocab = Vocabulary(method = method,
+                                use_pad = use_pad,
+                                use_unk = use_unk)
+
+    # Wrapper functions for vocabulary class,
+    # so that vocab is not directly exposed to
+    # outside user.
+    def size(self):
+        return self.__vocab.size()
+
+    def contains(self, entry: Any):
+        return self.__vocab.contians(entry)
+
+    def items(self):
+        return self.__vocab.items()
+
+    def add_entry(self, entry):
+        self.__vocab.add_entry(entry)
+
+    def entry2id(self, entry):
+        return self.__vocab.entry2id(entry)
+
+    def id2entry(self, idx):
+        return self.__vocab.id2entry(idx)
+
+    def get_pad_id(self):
+        return self.__vocab.get_pad_id()
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
+        raise NotImplementedError()
+
+    def extract(self, pack: DataPack, 
+            instance: EntryType) -> Feature:
+        raise NotImplementedError()
+
+    def add_to_pack(self, pack: DataPack, instance: EntryType, feature: Feature):
+        raise NotImplementedError()
+
+
+class AttributeExtractor(BaseExtractor):
+    def __init__(self, config: Dict):
+        super().__init__(config)
+        self.attribute = config["attribute"]
+
+    def update_vocab(self, pack: DataPack, instance: EntryType):
```

Comment:
Instance is different from attribute.
For example, the attributes could probably be a string, like "text", which will be used by getattr on the Annotation.
The instance is actually a specific Annotation in a specific place in the pack. Eg. it can be the second Sentence in the current pack. This instance need to be passed in each time, It tells the extractor, which Sentence is actually being processed right now

Reply:
This instance is a specific range in the pack. It is not an Annotation class but instead, it is an Annotation instance.